crates/noirc_evaluator/src/ssa/opt/die.rs

```diff
@@ -92,12 +92,26 @@ impl Context {
 
         let instruction = &function.dfg[instruction_id];
 
-        // These instruction types cannot be removed
-        if matches!(
-            instruction,
-            Constrain(_) | Call { .. } | Store { .. } | EnableSideEffects { .. }
-        ) {
-            return false;
+        match instruction {
+            // These instruction types can be safely removed
+            Binary(_)
+            | Cast(_, _)
+            | Not(_)
+            | Truncate { .. }
+            | Allocate
+            | Load { .. }
+            | ArrayGet { .. }
+            | ArraySet { .. } => (),
+
+            // These instruction types cannot be removed
+            Constrain(_) | Store { .. } | EnableSideEffects { .. } => return false,
```

Comment:
Since this match is larger now we should move it to a `has_side_effects` method on Instruction

Reply:
I thought about this but we'd need to pass in the `dfg`. That felt like a bit of an abstraction leak but is that acceptable?

Reply:
I think it's fine. We don't intend these instructions to be used outside of the dfg anyway.